### PR TITLE
Refined QDQ recipes of BERT/CLIP/VIT for QC and AMD.

### PR DESCRIPTION
**Pull request by tezheng**
## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [x] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: examples/clip/qdq/laion_clip_text_b32_qdq.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
+{
+    "input_model": {
+        "type": "PytorchModel",
+        "model_path": "laion/clip-vit-b-32-laion2b-s34b-b79k",
+        "generative": false,
+        "io_config": {
+            "input_names": [ "input_ids", "attention_mask" ],
+            "input_shapes": [ [ 1, 77 ], [ 1, 1, 77, 77 ] ],
```
Comment (jambayk) [inline examples/clip/qdq/laion_clip_text_b32_qdq.json:8]: Is the custom loader added to supported 4d mask and improve accuracy? We are expecting a PR soon which does some updates on the update attention mask surgery and makes it work for clip models as well. This is the commit if you would like to try it https://github.com/microsoft/Olive/commit/88a6a0b24f16a818725a3d957fd29e26b218a413. With this change, all ops can be quantized. 

If the accuracy is good with this change and latency is similar, I think using the default 2d mask and loader like the old examples might be better.

Comment (jambayk) [reply examples/clip/qdq/laion_clip_text_b32_qdq.json:8]: #1803

File: examples/bert/qdq/google_bert_base_qdq.json
[the remark this answers is not in the document]
Comment (jambayk) [reply examples/bert/qdq/google_bert_base_qdq.json:1]: bumping this again. Are we keeping both the old and new qdq configs?

File: examples/clip/qdq/eval_retrieval_qnn.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    result = eval_retrieval_accuracy(
+        args.model,
+        args.text_encoder,
+        args.image_encoder,
+        args.dataset,
+        args.split,
+        args.tokenizer,
+    )
+
+    print("Evaluation result", format_output(result))
```
[the remark this answers is not in the document]
Comment (tezheng) [reply examples/clip/qdq/eval_retrieval_qnn.py:301]: This is inference sample code, can we ignore the warning?

**Comment by tezheng** [comment]
@jambayk I think the lint errors are false alarms, do you have any ideas/suggestions?

File: examples/bert/qdq/google_bert_base_qdq.json
[the remark this answers is not in the document]
Comment (jambayk) [reply examples/bert/qdq/google_bert_base_qdq.json:1]: @tezheng

**Comment by jambayk** [comment]
> @jambayk I think the lint errors are false alarms, do you have any ideas/suggestions?

There were some lint rules and requirements update on main. Please try `lintrunner init` to use the latest. For this PRm I have pushed a commit with the lint fixes.

File: examples/bert/qdq/google_bert_script.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def load_bert_nsp_model(model_name: str) -> torch.nn.Module:
+    return AutoModelForNextSentencePrediction.from_pretrained(model_name).eval()
+
+
+@Registry.register_post_process()
+def bert_scl_post_process(outputs):
+    """Post-processing for Sequence Classification task."""
+    match outputs:
```
Comment (jambayk) [inline examples/bert/qdq/google_bert_script.py:30]: Olive still supports python 3.9 so this syntax is invalid. lint fails because of this.

File: examples/bert/qdq/google_bert_base_qdq.json [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            "dim_value": [ 1, 512 ]
+        },
+        "surgery": {
+            "type": "GraphSurgeries",
+            "surgeries": [
+                { "surgeon": "ReplaceAttentionMaskValue", "replacement": -100.0 },
+                { "surgeon": "MatMulAddToGemm" }
+            ]
+        },
+        "transformer_optimizer": {
```
Comment (jambayk) [inline examples/bert/qdq/google_bert_base_qdq.json:28]: I think this might not be required? with opset 20, the model gets exported with layernorm operator and quant_preprocess in onnxstatic quantization performs gelu fusion automatically

File: examples/bert/qdq/google_bert_base_qdq.json [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        "to_fixed_shape": {
+            "type": "DynamicToFixedShape",
+            "dim_param": [ "batch_size", "sequence_length" ],
+            "dim_value": [ 1, 512 ]
+        },
+        "surgery": {
+            "type": "GraphSurgeries",
+            "surgeries": [
+                { "surgeon": "ReplaceAttentionMaskValue", "replacement": -100.0 },
+                { "surgeon": "MatMulAddToGemm" }
```
Comment (jambayk) [inline examples/bert/qdq/google_bert_base_qdq.json:25]: this works well for qc. not sure if it's recommended for amd.

File: examples/bert/qdq/google_bert_base_qdq.json [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            "input_names": [ "input_ids", "attention_mask", "token_type_ids" ],
+            "input_shapes": [ [ 1, 512 ], [ 1, 512 ], [ 1, 512 ] ],
+            "input_types": [ "int32", "int32", "int32" ],
+            "output_names": [ "logits" ]
+        },
+        "model_loader": "load_bert_nsp_model",
+        "model_script": "google_bert_script.py"
+    },
+    "passes": {
+        "conversion": { "type": "OnnxConversion", "target_opset": 20, "dynamic": true, "use_dynamo_exporter": false },
```
Comment (jambayk) [inline examples/bert/qdq/google_bert_base_qdq.json:15]: dynamic and use_dynamo_exporter options are not required

File: examples/bert/qdq/wikitext.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        data = {
+            sent_cols[0]: [s1 for s1, _ in pairs],
+            sent_cols[1]: [s2 for _, s2 in pairs],
+            label_col: labels,
+        }
+
+    return Dataset.from_pandas(pd.DataFrame(data))
+
+
+def parse_args():
```
Comment (jambayk) [inline examples/bert/qdq/wikitext.py:78]: do we need this and the __main__ part?

File: examples/clip/qdq/openai_clip_text_b16_qdq.json [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                { "surgeon": "ReplaceAttentionMaskValue", "replacement": -100.0 },
+                { "surgeon": "MatMulAddToGemm" }
+            ]
+        },
+        "transformer_optimizer": {
+            "type": "OrtTransformersOptimization",
+            "model_type": "bert",
+            "opt_level": 1,
+            "optimization_options": {
+                "enable_gelu": false,
```
Comment (xieofxie) [inline examples/clip/qdq/openai_clip_text_b16_qdq.json:34]: this will not compile on AMD NPU

File: examples/bert/qdq/google_bert_base_qdq.json
[the remark this answers is not in the document]
Comment (xieofxie) [reply examples/bert/qdq/google_bert_base_qdq.json:1]: We have already released parts of these in AITK and I created a pr https://github.com/microsoft/Olive/pull/1893 to align the recipes. It is part of this pr: CLIP text + vision, bert and vit qdq.
Please take a look @jambayk  CC @tezheng 

I updated the previous qdq configs directly

[the remark this answers is not in the document]
Comment (xieofxie) [reply examples/bert/qdq/google_bert_base_qdq.json:1]: For others, we could take time to merge if feel needed

**Comment by jambayk** [comment]
closing this is it has gone stale and new PRs related to this has merged or are open.
